### PR TITLE
166 plotting with correct projection config using pcolormesh

### PR DESCRIPTION
**Pull request by sadamov**
## Describe your changes

Switched to latitude/longitude based plotting using `pcolormesh` and `cartopy` for accurate spatial visualization. This method comes directly from the research branch.

- **Lat/Lon Plotting with `pcolormesh`**: Updated `neural_lam/vis.py` to use `ax.pcolormesh` with `ccrs.PlateCarree()` transform, using explicit latitude and longitude coordinates from the datastore. This ensures that data is plotted correctly on the map regardless of the underlying projection, fixing previous misalignment issues.
- **LatLon Retrieval**: 
    - **Base Implementation**: The default `get_lat_lon` in `BaseRegularGridDatastore` derives latitude/longitude from the Cartesian `x`/`y` coordinates using the defined projection (`coords_projection`).
    - **MDP Override**: The new implementation in `MDPDatastore` first checks if explicit latitude and longitude coordinates (e.g., `lat`/`lon` or `latitude`/`longitude`) exist in the dataset. If found, it uses these directly, bypassing the projection transformation.
- **Visualization Enhancements**:
    - Added `boundary_alpha` argument to overlay a semi-transparent mask on boundary regions.
    - Added `crop_to_interior` argument to automatically crop the plot view to the valid interior region.
    Note: The boundary vizualisation is deemed to change soon, with the separate boundary-datastore inclusion.

## Issue Link

solves #166

## Figures

Here are two examples for MEPS and DANRA test datastores with their own projections (`boundary_alpha = 0.7, crop_to_interior = True`).

### These are the spatial loss maps, generated during val and test

Screenshot From 2025-11-20 10-32-16
Screenshot From 2025-11-20 10-14-19

### These are the predictions generated during test

image
image

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [x] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by sadamov** [comment]
@lorenzo30salgado I am tagging you here, because I couldn't add you as a reviewer.
@leifdenby Let me know if these extensions to the datastores are what you had in mind as well.
@joeloskarsson do we want to keep the spatial loss maps so minimal without any titles/labels?

**Comment by joeloskarsson** [comment]
We could add some more info to the loss maps, I have no strong opinions. Maybe title and lead times could be nice to add.

File: neural_lam/datastore/base.py [hunk 1]
```diff
@@ -455,6 +472,9 @@ class BaseRegularGridDatastore(BaseDatastore):
     - `get_xy` (method): Return the x, y coordinates of the dataset, with the
       option to not stack the coordinates (so that they are returned as a 2D
       grid).
+        - `get_lat_lon` (method): Return the latitude/longitude coordinates of
+          the
+            dataset for convenience when plotting.
```
Comment (joeloskarsson) [inline neural_lam/datastore/base.py]: Just some weird indentation
```suggestion
    - `get_lat_lon` (method): Return the latitude/longitude coordinates of
      the dataset for convenience when plotting.
```

File: neural_lam/vis.py [hunk 2]
```diff
@@ -67,111 +193,102 @@ def plot_error_map(errors, datastore: BaseRegularGridDatastore, title=None):
 @matplotlib.rc_context(utils.fractional_plot_bundle(1))
 def plot_prediction(
     datastore: BaseRegularGridDatastore,
-    da_prediction: xr.DataArray,
-    da_target: xr.DataArray,
+    da_prediction: xr.DataArray = None,
+    da_target: xr.DataArray = None,
```
Comment (joeloskarsson) [inline neural_lam/vis.py]: Is case where one of these is `None` every properly handled below? Otherwise we should avoid to have that as a possible default.

Comment (joeloskarsson) [reply neural_lam/vis.py]: Maybe just check for `None` before calling `plot_on_axis` for each?

File: neural_lam/vis.py [hunk 3]
```diff
@@ -67,111 +193,102 @@ def plot_error_map(errors, datastore: BaseRegularGridDatastore, title=None):
...
-    # Get common scale for values
+    """Plot spatial error with projection-aware axes."""
+
+    grid_shape = [datastore.grid_shape_state.x, datastore.grid_shape_state.y]
+
+    value_source = error
+    if hasattr(value_source, "detach"):
+        value_source = value_source.detach()
+    if hasattr(value_source, "cpu"):
+        value_source = value_source.cpu()
```
Comment (joeloskarsson) [inline neural_lam/vis.py]: If we assume that this is tensor (and add that as type annotation in the function), we could just always do `.detach().cpu()`.

File: neural_lam/vis.py [hunk 4]
```diff
@@ -9,6 +12,129 @@
...
+    crop_to_interior=False,
+):
+    """Plot weather state on given axis using datastore metadata.
+
+    Parameters
+    ----------
+    ax : matplotlib.axes.Axes
+        The axis to plot on. Should have a cartopy projection.
+    da : xarray.DataArray or np.ndarray
+        The data to plot. Should have shape (N_grid,).
```
Comment (joeloskarsson) [inline neural_lam/vis.py]: Is it a good idea (from a maintainability standpoint) to allow this to be either `xarray.DataArray` or `np.ndarray`? I am leaning towards just enforcing this to be an `xarray.DataArray`. I think that would simplify some parts of this function, and I believe all calls to this anyhow use dataarrays now.

File: neural_lam/datastore/base.py (hunk 1 above)
Comment (sadamov) [reply neural_lam/datastore/base.py]: Fixed in https://github.com/mllam/neural-lam/pull/168/commits/66decb964a600fa4411e964b4c0922e0a6a6504d

File: neural_lam/vis.py (hunk 3 above)
Comment (sadamov) [reply neural_lam/vis.py]: smart, I added this stricter type annotation as suggested in https://github.com/mllam/neural-lam/pull/168/commits/47fbdf86887020a17acef7d9e81813f61cb82d15

File: neural_lam/vis.py (hunk 2 above)
Comment (sadamov) [reply neural_lam/vis.py]: I removed the None from the types, none-related errors are now handled by xarrays default error handling. see https://github.com/mllam/neural-lam/pull/168/commits/47fbdf86887020a17acef7d9e81813f61cb82d15

File: neural_lam/vis.py (hunk 4 above)
Comment (sadamov) [reply neural_lam/vis.py]: also here I followed the same design principle, stricter types -> easier maintainability. https://github.com/mllam/neural-lam/pull/168/commits/47fbdf86887020a17acef7d9e81813f61cb82d15

**Comment by sadamov** [comment]
Hi @joeloskarsson I finally found some time to react to your review. I agree with all your points and made the static typing stricter for easier maintainability. Since it has been a while and we have seen many developments recently, there is some additional code that fits into this PR, and needs additional review. Specifically:
- There are new tests contributed @Jayant-kernel https://github.com/mllam/neural-lam/pull/168/commits/3bb423670dfea2ad5f39ddce8989049a6ca4dee1
- I added a little helper function to plot strange symbols in tex mode (e.g. % unit without escape): https://github.com/mllam/neural-lam/pull/168/commits/6a2895d1f8bbe2cf4b9d15cffef848ed0efd33d6
- Here I introduce some constants to get identical layouts and labels for both the error and prediction maps: https://github.com/mllam/neural-lam/pull/168/commits/6655d2cc7ef49af66f0cf6cc67cdc77bb0c8d2f9
  - the unit is displayed beneath the colorbar now
  - I was too eager to define the plot layout myself, so here I reverted unnecessary manual implementation back to the rc_bundle: https://github.com/mllam/neural-lam/pull/168/commits/da67fce7d87cf60cbcff90c4ba3cee5ec6054973

The result look like this:

image
image


Please review again and let me know if some of these additional changes should become a separate PR.

**Comment by sadamov** [comment]
yes that was just my env not having tex installed. merging :raised_hands:
